packages/scripts/src/clusterize.ts

```diff
@@ -0,0 +1,89 @@
+import dbscan from "@cdxoo/dbscan";
+import { strict as assert } from "assert";
+
+/**
+  Based on https://www.npmjs.com/package/@cdxoo/dbscan
+ */
+export type DbscanOptions = {
+  /**
+    Maximum distance between datapoints. Determine if a datapoint is in a
+    cluster or not. Default is 1.0.
+   */
+  epsilon: number;
+
+  /**
+    Custom function to compare calculated distance and epsilon. Default is
+    (dist, e) => (dist < e)
+   */
+  epsilonCompare: (distance: number, epsilon: number) => boolean;
+
+  /**
+    Threshold of how many points are needed in the same neighborhood to form a
+    cluster. Default is 2.
+   */
+  minimumPoints: number;
+};
+
+const defaultDbscanOptions: DbscanOptions = {
+  epsilon: 1.0,
+  epsilonCompare: (d, e) => d < e,
+  minimumPoints: 2,
+};
+
+/**
+  Groups arbitrary data into clusters.
+  @param dataset The collection of instances to group into clusters.
+  @param getVector A function that, given an element of the dataset, returns the
+  spatial position of that element.
+  @param optionsIn Additional dbscan options.
+ */
+export const clusterize = <T>(
+  dataset: T[],
+  getVector: (x: T) => number[],
+  optionsIn?: Partial<DbscanOptions>
+): {
+  /**
+    Each cluster is represented as an array of the instances that fell into that
+    cluster.
+   */
+  clusters: T[][];
+
+  /**
+    The "noise" array contains the instances that did not fall into any cluster.
+   */
+  noise: T[];
+} => {
+  if (dataset.length === 0) {
+    return { clusters: [], noise: [] };
+  }
+
+  const options: DbscanOptions = {
+    ...defaultDbscanOptions,
+    ...(optionsIn ?? {}),
+  };
+
+  const { clusters: clusterIndexesArray, noise: noiseIndexes } = dbscan({
+    ...options,
+    dataset,
+    distanceFunction(a, b) {
+      return distanceSquared(getVector(a), getVector(b));
```

Comment:
This is a Euclidean distance metric right?

My (limited) understanding is that in a many-dimensional space the Euclidean metric distance between _any_ two given vectors is ~1. i.e. the usefulness of the Euclidean metric is inversely proportional to number of dimensions

Should we use cosine similarity instead?

Reply:
Maybe? Check the results in the dev database, seems to do a pretty good job at distinguishing different questions and grouping similar questions with a reasonable epsilon value. Open to suggestions though.

Reply:
Not a blocker here, especially if our results are satisfying. Maybe something to experiment with in future!